android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java

```diff
@@ -93,8 +96,33 @@
 	private TiWeakList<OnPrepareOptionsMenuEvent> onPrepareOptionsMenuListeners = new TiWeakList<OnPrepareOptionsMenuEvent>();
 	private APSAnalytics analytics = APSAnalytics.getInstance();
 
-	public static KrollObject storageCallbackContext, cameraCallbackContext, contactsCallbackContext, oldCalendarCallbackContext, calendarCallbackContext, locationCallbackContext;
-	public static KrollFunction storagePermissionCallback, cameraPermissionCallback, contactsPermissionCallback, oldCalendarPermissionCallback, calendarPermissionCallback, locationPermissionCallback;
+
+	public static class PermissionContextData {
+		private final Integer requestCode;
+		private final KrollObject context;
+		private final KrollFunction callback;
+
+		public PermissionContextData(Integer requestCode, KrollFunction callback,
+									 KrollObject context) {
+			this.requestCode = requestCode;
+			this.callback = callback;
+			this.context = context;
+		}
+
+		public Integer getRequestCode() {
+			return requestCode;
+		}
+
+		public KrollFunction getCallback() {
+			return callback;
+		}
+
+		public KrollObject getContext() {
+			return context;
+		}
+	}
+
+	private static ConcurrentHashMap<Integer,PermissionContextData> callbackDataByPermission = new ConcurrentHashMap<Integer, PermissionContextData>();
```

Comment:
I did a check and there was no leak on WindowProxy by opening 10 windows and requesting (and denying) permissions using the code below:-

```javascript
(function openWindow(n) {
    var win = Ti.UI.createWindow({
        backgroundColor : "#aaaaaa",
        layout : "vertical"
    });
 
    win.title = "Window # " + n;
 
    var label = Ti.UI.createLabel({
        text : "Window # " + n,
        color : "#ffffff",
        top : 20
    });
 
    win.add(label);
    label = null;
 
    var openButton = Ti.UI.createButton({
        title : "Open",
        color : "#ffffff",
        top : 20,
        _n: n
    });
 
    openButton.addEventListener('click', function onOpenByButton(evt) {
        openWindow(evt.source._n + 1);
    });
 
    win.add(openButton);
    openButton = null;
 
    var closeButton = Ti.UI.createButton({
        title : "Close",
        color : "#ffffff",
        top : 20
    });
 
    closeButton.addEventListener('click', function onCloseByButton(evt) {
        evt.source.getParent().close();
    });
 
    win.add(closeButton);

    win.open();
    Ti.Media.requestCameraPermissions(function(e) {
        if (e.success === true) {
                Ti.API.log("Success");
            } else {
                Ti.API.log("Access denied, error: " + e.error);
            }
    });
    win = null;
})(1);
```
Code is based on https://jira.appcelerator.org/browse/TIMOB-19891. 

No leaks seen when hprof was inspected for WindowProxy.